### PR TITLE
Expand swap capability and reorganize package

### PR DESCRIPTION
**Pull request by zietzm**
In trying to apply the XSwap method to other networks, I have discovered that this implementation fails for several publicly-available networks. A few improvements address most of the underlying issues.

First, I added a preprocessing file to process files containing graphs. The underlying C++ functionality deals with edges represented as `int*`s. Since permutable graphs may be in files as integers or strings, (eg: `node_a,node_b\nnode_c,node_d\n`...), it is helpful to be able to process such graphs to a form that is usable by the fast XSwap implementation. Moreover, assigning new ids to the nodes allows us to ensure that nodes are indexed 1, 2, ..., `num_edges`. This kind of indexing allows larger graphs to be memory-feasible using the fast Cantor pair hash table.

Second, I added an additional hash table implementation using the C++ standard library `std::unordered_set`. While this is significantly slower than the Cantor pair hash table, it will work for networks with too many potential edges for the original hash table to fit in memory. (I have set the cutoff at 4 GB, though this was a totally arbitrary choice.) To ensure a uniform API, I added a final hash table wrapper, the `HashTable` class, which can call both the original `EdgeHashTable` and the new `BigHashTable`.

### CODE REVIEW

**Comment by zietzm** [comment]
As an example, using Zachary's karate club network with `multiplier=10000`, we find that it does not affect the swap outcomes, but significantly affects computation time.

Using slow hash table
43.5897 % unchanged
5.205 seconds

Using fast hash table
43.5897 % unchanged
0.134 seconds

**Comment by dhimmel** [comment]
In the previous comment, does "fast hash table" refer to the "Cantor pair hash table" and "slow hash table" refer to the `std::unordered_set` hash table?

File: xswap/preprocessing.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
+import pathlib as _pathlib
```
Comment (dhimmel) [inline xswap/preprocessing.py]: Why the `_pathlib`? I'm guessing so it doesn't show up in the namespace to external modules? Although I don't usually see this done for module imports

File: xswap/preprocessing.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    as Tuple[str, str]. Used to load edges for preprocessing.
+    """
+    file_path = _pathlib.Path(filename)
+    if not file_path.is_file():
+        raise FileNotFoundError
+
+    with open(file_path, 'r') as f:
+        graph_string = f.read()
+    str_edges = [
+        edge.split(node_delim) for edge in graph_string.split(edge_delim) if edge != ''
```
Comment (dhimmel) [inline xswap/preprocessing.py]: Would it not make sense to use Python's CSV module here, which could deal with things like quoting.

File: xswap/preprocessing.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    """
+    Return a dict mapping a list of nodes to their sorted indices. Nodes should
+    be a list of strings.
+
+    Returns:
+    --------
+    Dict[str, int]
+    """
+    name_to_id = dict()
+    sorted_node_set = sorted(list(set(nodes)))
```
Comment (dhimmel) [inline xswap/preprocessing.py]: list can be removed here. sorted returns a list

File: xswap/preprocessing.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    for i, name in enumerate(sorted_node_set):
+        name_to_id[name] = i
```
Comment (dhimmel) [inline xswap/preprocessing.py]: Can be done in a single line with dict comprehension:

```python
name_to_id = {name: i for i, name in enumerate(sorted_node_set)}
```

**Comment by zietzm** [comment]
That is basically correct, though I have found that `std::set` actually gives better performance than `std::unordered_set`. I am not sure why, but `std::unordered_set` had been giving insertion performance much closer to the worst case (O(n)) than to the best case (constant). Since `std::set` uses a self-balancing binary search tree (O(log(n))), log(n) is proving to be faster than the worst-case hash table. I am looking into the possibility that a better, custom-specified hash function would improve performance on the hash table.

Here's a quick rundown of the differences in implementations between `set` and `unordered_set`:
https://www.geeksforgeeks.org/set-vs-unordered_set-c-stl/

File: xswap/preprocessing.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    return name_to_id
+
+
+def _apply_map(edges, mapping):
+    """
+    Maps edges according to a node mapping. If `mapping` contains two dicts,
+    `_apply_map` maps source and target nodes separately. Else, if `mapping`
+    contains only one dict, then source and target nodes are mapped according
+    to the same mapping.
+    """
```
Comment (dhimmel) [inline xswap/preprocessing.py]: This seems a little hacky. Would it make sense for the function to take both source_map and target_map? Then the calling function could just pass the same dict to source_map and target_map in the case they are of the same metanode.

**Comment by zietzm** [comment]
[cd5fea4e0958f32812f1919109f798fea6c1b82f](https://github.com/greenelab/xswap/pull/6/commits/cd5fea4e0958f32812f1919109f798fea6c1b82f) closes #7

**Pull request merged**